README.md

```diff
@@ -0,0 +1,113 @@
+# KCP Helm Charts
+
+Repository for KCP helm charts.
+
+## Pre-requisites
+
+- Cert-manager installed and running
+- Ingress installed (e.g. nginx-ingress or OpenShift router)
+
+## Usage
+
+[Helm](https://helm.sh) must be installed to use the charts.  Please refer to
+Helm's [documentation](https://helm.sh/docs) to get started.
+
+Once Helm has been set up correctly, add the repo as follows:
+
+    ```bash
+    helm repo add kcp https://kcp-dev.github.io/helm-charts
+    ```
+
+If you had already added this repo earlier, run `helm repo update` to retrieve the latest versions of the packages.  You can then run `helm search repo kcp` to see the charts.
+
+To install the kcp chart:
+
+    ```bash
+    helm install my-kcp kcp/kcp
+    ```
+
+To uninstall the chart:
+
+    ```bash
+    helm delete my-kcp
+    ```
+
+## Development usage
+
+To install using the local chart:
+
+    ```bash
+    helm install kcp ./charts/kcp --values ./myvalues.yaml --namespace kcp --create-namespace
+    ```
+
+Changes can then be made locally and tested via upgrade:
+
+    ```bash
+    helm upgrade kcp ./charts/kcp --values ./myvalues.yaml --namespace kcp
+    ```
+
+Note `myvalues.yaml` will depend on your environment (you must specify which ingress method
+is used to expose the front-proxy endpoint), a minimal example:
+
+    ```yaml
+    externalHostname: "<external hostname as exposed by ingress method below>"
+    kcpFrontProxy:
+      ingress:
+        enabled: true
+        annotations:
+          kubernetes.io/ingress.class: "nginx"
+          nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
+      certificates:
+        issuerSpec:
+          acme:
+            server: https://acme-v02.api.letsencrypt.org/directory
+            privateKeySecretRef:
+              name: kcp-front-proxy-issuer-account-key
+            solvers:
+            - http01:
+                ingress:
+                  serviceType: ClusterIP
+    ```
+### Install to KIND cluster (for development)
+
+There is a helper script to install KCP to a KIND cluster. It will install cert-manager, nginx-ingress and KCP.
+Kind cluster binds to host ports 80 and 443 for ingress. Ingress is emulated using host entries in `/etc/hosts`.
+This particular configuration is useful for development and testing, but will not work with LetsEncrypt.
+
+    ```bash
+    ./hack/kind-setup.sh
```

Comment:
This fails for me like so:

```
❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Thanks for using kind! 😊
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
Error from server (NotFound): nodes "faros-control-plane" not found
```

Reply:
With this fix:
```patch
diff --git a/hack/kind-setup.sh b/hack/kind-setup.sh
index 93fd8f1..ac179ad 100755
--- a/hack/kind-setup.sh
+++ b/hack/kind-setup.sh
@@ -25,7 +25,7 @@ export KUBECONFIG=./${CLUSTER_NAME}.kubeconfig
 echo "Installing ingress"
 
 kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/kind/deploy.yaml
-kubectl label nodes faros-control-plane node-role.kubernetes.io/control-plane-
+kubectl label nodes kcp-control-plane node-role.kubernetes.io/control-plane-
 
 echo "Waiting for the ingress controller to become ready..."
 kubectl --context "${KUBECTL_CONTEXT}" -n ingress-nginx wait --for=condition=Ready pod -l app.kubernetes.io/component=controller --timeout=5m
```
I get a little further, but it stills fails eventually:
```
❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
node/kcp-control-plane unlabeled
Waiting for the ingress controller to become ready...
error: no matching resources found

```

Reply:
Still:
```
❯ git rev-parse @
58d8622466f304bc02ee7fdcaef39b45f8c7df09

kcp-helm-charts  local.dev [$]  ⊙ 4

❯ ./hack/kind-setup.sh
KIND already installed
No kind clusters found.
Creating cluster "kcp" ...
 ✓ Ensuring node image (kindest/node:v1.24.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kcp"
You can now use your cluster with:

kubectl cluster-info --context kind-kcp --kubeconfig ./kcp.kubeconfig

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
Installing ingress
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
node/kcp-control-plane unlabeled
Waiting for the ingress controller to become ready...
error: no matching resources found
```

Reply:
This implies ingress is not deploying for you. Can you check the cluster to see what is happening there?

Reply:
`ingress-nginx` is deploying fine, but there's a race condition.

The installation of `ingress-nginx` and waiting for its readiness happens immediately after the kind cluster is created. At that time the control plane-related pods within the cluster haven't even started yet. As a result, at that moment it seems as if `ingress-nginx` is not installing.

If you bump the sleep to `30` it works, because everything had a chance to wake up. At least on my machine. `sleep 10` was too soon as well.

I don't think that sleeping here is the right thing. Maybe decouple cluster creation and initialization into different scripts. Or, better, find a more resilient and idiomatic approach to waiting.

Reply:
I pushed some different changes. 
Moving to a separate script IMHO will not solve anything as we still have the same issue, just in the other script.

Reply:
I'm afk for some days. I've seen the changes you and assume it's working. Thanks for heeding my feedback!